### ISSUE TITLE
java.net.ConnectException: Connection refused error on 0.16 version

### ISSUE THREAD

**Issue by land1725** [Azure/mmlspark#536]
Hello,I am so sorry to tell that I met java.net.ConnectException on 0.16 version 

> 
        <dependency>
            <groupId>Azure</groupId>
            <artifactId>mmlspark</artifactId>
            <version>0.16</version>
        </dependency>
this is my command to start spark
spark2-submit --class  HuoyingLightGBM --master yarn **--executor-memory 8G --num-executors 2 --driver-memory 8g --executor-cores 1**  --conf spark.yarn.executor.memoryOverhead=1024 --conf spark.sql.shuffle.partitions=20 /data/recommend.jar

this is my code:

    val executor = ss.sparkContext.getConf.getInt("spark.executor.instances", 2)
    println("lucky spark.executor.instances is " + executor)**//this is 2**
    val finamInput = trainInput.repartition(executor)

    val classifierModel = new LightGBMClassifier()
      .setLabelCol("label")
      .setFeaturesCol("features")
      .setRawPredictionCol("rawPrediction")
      .setObjective("binary")
      //      .setParallelism("voting_parallel")
      .setLearningRate(0.1)
      .setMaxBin(128)
      .setNumLeaves(500)
      .setMaxDepth(10)
      .setFeatureFraction(0.8)
      .setNumIterations(10)
      .setBaggingFraction(1.0)
      .setBaggingFreq(5)
      .setDefaultListenPort(24001)
      .setTimeout(100)
      .fit(finamInput)
    
    val testRank = classifierModel.transform(testInput)
    testRank.show(100, false)
    val binaryClassificationEvaluator = new BinaryClassificationEvaluator()
      .setMetricName("areaUnderROC")
      .setRawPredictionCol("rawPrediction")
      .setLabelCol("label")
    val classificationMetric = binaryClassificationEvaluator.evaluate(testRank)

    val epsilon = 1e-8
    println(s"LightGBMClassifier ${binaryClassificationEvaluator.getMetricName} = $classificationMetric")

and this is driver and worker logs.

[worker1.txt](https://github.com/Azure/mmlspark/files/3062423/worker1.txt)
[worker2.txt](https://github.com/Azure/mmlspark/files/3062424/worker2.txt)
[driver.txt](https://github.com/Azure/mmlspark/files/3062435/driver.txt)

**Comment by imatiach-msft** [comment]
@land1725 I looked at the logs but didn't see any error message that might help me understand what is causing the errors.  The connection refused errors could be due to a number of issues.  For the driver logs, you sent me one file but usually there are three files which contain different debug statements: 1.) stdout 2.) stderr 3.) log4j logs
Would you be able to send me the other driver logs?

**Comment by land1725** [comment]
@imatiach-msft hello,the driver log has already contains 1.) stdout 2.) stderr 3.) log4j logs

**Comment by imatiach-msft** [comment]
@land1725 hmm, perhaps it does have stderr/stdout, I'm not sure, but I do see the log4j logs.
I never see this line printed however:
https://github.com/Azure/mmlspark/blob/0b84a230d1556ced87be9139dd798237711c1158/src/lightgbm/src/main/scala/LightGBMUtils.scala#L115
It seems the workers aren't able to make a socket connection to the driver.  I wonder why that is.

**Comment by imatiach-msft** [comment]
I do see this line in the driver log:
19/04/10 14:48:09 INFO spark.LightGBMClassifier: driver waiting for connections on host: 10.111.1.207 and port: 41185
So it seems like the driver is waiting on this host/port but the workers are not able to connect to it.
But as to why the connection is refused from the workers to the driver I'm not quite sure.

**Comment by land1725** [comment]
@imatiach-msft hello,how can I help you to find how to solve this problem.Do you need to add more Log Info and Build a new PR for me?

**Comment by imatiach-msft** [comment]
@land1725 might you have some cluster settings that prevent the workers from connecting to the driver node?  Did lightgbm work on this cluster in an earlier version of mmlspark before?

**Comment by land1725** [comment]
@imatiach-msft it is not possible about cluster settings that prevent the workers from connecting to the driver node.Because I use LightGBMRegressor,it is fine.
my mmlspark version is  latest 0.16         
<groupId>Azure</groupId>
<artifactId>mmlspark</artifactId>
<version>0.16</version>

**Comment by imatiach-msft** [comment]
you can try to run some simple scala code in a notebook against the cluster.  For example, you can do something similar to method createDriverNodesThread:
https://github.com/Azure/mmlspark/blob/0b84a230d1556ced87be9139dd798237711c1158/src/lightgbm/src/main/scala/LightGBMUtils.scala#L97
specifically, create a server socket:
val driverServerSocket = new ServerSocket(0)
and then get host and port and wait for workers to connect:
val host = getDriverHost(df)
val port = driverServerSocket.getLocalPort
Then, do a mapPartitions and verify that the workers can connect to the driver using those settings.
One strange thing I see in the logs is that your host is 10.111.1.204:
ApplicationMaster host: 10.111.1.204
However, in the logs it says:
driver waiting for connections on host: 10.111.1.207 and port: 41185
It seems strange that those hosts are different?

**Comment by imatiach-msft** [comment]
@land1725 another useful thing may be to go to the yarn UI, go to the failing task and you should see an error for each worker.  However, it looks like you have already given me the logs for both workers and they had the same error (connection refused).  I am a bit surprised because usually that error is a red herring and the real error is on the other worker.  Were those logs from the 2 workers for the same run?

**Comment by imatiach-msft** [comment]
@land1725 if LightGBMRegressor works fine than something is very strange, because both regressor and classifier use the same method to connect the workers to driver.  The differences only appear after the initial connection has been established.

**Comment by land1725** [comment]
@imatiach-msft I have did a simple test ,it seems good
scala>     val host = getDriverHost(df)
host: String = 10.111.1.207

scala> val port = driverServerSocket.getLocalPort
port: Int = 40678

scala>     val port = driverServerSocket.getLocalPort
port: Int = 40678

scala>     log.info(s"driver waiting for connections on host: ${host} and port: $port")

scala> df.repartition(2)
res21: org.apache.spark.sql.Dataset[org.apache.spark.sql.Row] = [vid: int, features: array<float>]

scala> res21.rdd.count
res22: Long = 410040

**Comment by imatiach-msft** [comment]
@land1725 one thing you could try is call mapPartitions on the df, and inside it try to connect to the driver from the workers, similar to the code here:
https://github.com/Azure/mmlspark/blob/master/src/lightgbm/src/main/scala/TrainUtils.scala#L242
and then send some info like here:
https://github.com/Azure/mmlspark/blob/master/src/lightgbm/src/main/scala/TrainUtils.scala#L250
the workers should be able to connect to the driver using the driver's host and port and send some information over the socket.

**Comment by imatiach-msft** [comment]
@land1725 one thing I noticed, you are setting the port to be:
```
  .setDefaultListenPort(24001)
```
could you leave it as the default value:
val defaultLocalListenPort = 12400
probably won't make any difference, but worth a try.

**Comment by land1725** [comment]
@imatiach-msft hello ,I am happy to tell that this problems has been solvec.
in My code you can see I set timeout 100s 
.setTimeout(100)
when I delete this code it is normal again,thanks
